---
title: "Raw data"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, results='hide', message=FALSE, warning=FALSE, fig.height = 7, fig.width = 7)
dyn.load('/software/geos-3.9.1-el8-x86_64/lib64/libgeos_c.so')
library(ggplot2)
library(dplyr)
library(plyr)
library(stringr)
library(harmony)
library(Seurat)
library(ArchR)
library(parallel)
library(BSgenome.Hsapiens.UCSC.hg38)
source('~/yuzhao1/scripts/plot.R')

```

```{r, include = F, eval=F}
source('~/yuzhao1/scripts/helper_archr.R')
source('/project/gca/yuzhao1/work/manu/rc2/scripts/rna_annotation_markers.R')
source('~/yuzhao1/work/manu/rc2/scripts/umap_colors.R')
source('~/yuzhao1/work/manu/rc2/scripts/rna_deg_markers.R')
```

```{r}
out.dir <- '~/yuzhao1/work/manu/rc2/plots/2rna_degs_volcano_others/'
```

```{r}
seurat <- readRDS('~/yuzhao1/work/final_RC2rna/annotation/rds/gut_filtered.rds')
```

```{r}
# markers_GB2vsGB1 <- FindMarkers(seurat, ident.1 = 'Goblet2', ident.2 = 'Goblet1', group.by = 'anno1', logfc.threshold = 0)
# saveRDS(markers_GB2vsGB1, paste0(out.dir, 'markers_GB2vsGB1.rds'))
```


```{r}
#higher resolution in goblet
seurat$anno1[seurat$anno1 %in% c("Goblet1") & seurat$biopsy_location =='POU'] <- 'Goblet-POU1'
seurat$anno1[seurat$anno1 %in% c("Goblet2") & seurat$biopsy_location =='POU'] <- 'Goblet-POU2'
seurat$anno1[seurat$anno1 %in% c("Goblet1", "Goblet2") & seurat$biopsy_location =='PP'] <- 'Goblet-PP'
seurat$anno1[seurat$anno1 %in% c("Goblet1", "Goblet2") & seurat$biopsy_location =='TI'] <- 'Goblet-TI'
seurat$anno1[seurat$anno1 %in% c("Goblet1", "Goblet2") & seurat$biopsy_location =='AC'] <- 'Goblet-AC'

POU_GB2vsAC <- FindMarkers(seurat, ident.1 = 'Goblet-POU2', ident.2 = 'Goblet-AC', group.by = 'anno1', logfc.threshold = 0)
saveRDS(POU_GB2vsAC, '~/yuzhao1/work/manu/rc2/plots/2rna_degs_volcano_others/POU_GB2vsAC.rds')

POU_GB2vsGB1 <- FindMarkers(seurat, ident.1 = 'Goblet-POU2', ident.2 = 'Goblet-POU1', group.by = 'anno1', logfc.threshold = 0)
saveRDS(POU_GB2vsGB1, '~/yuzhao1/work/manu/rc2/plots/2rna_degs_volcano_others/POU_GB2vsGB1.rds')



markers_GB2vsGB1 <- readRDS(paste0(out.dir, 'markers_GB2vsGB1.rds'))
markers_POU_GB2vsAC <- readRDS(paste0(out.dir, 'POU_GB2vsAC.rds'))
```

```{r}
# check #genes
markers <- readRDS('~/yuzhao1/work/manu/rc2/plots/2rna_degs_volcano_others/POU_GB2vsGB1.rds')
length(rownames(markers)[(markers$p_val < 0.001 & markers$avg_log2FC < -1)])
length(rownames(markers)[(markers$p_val < 0.001 & markers$avg_log2FC > 1)])

```


```{r}
plot.title = "Pouch Goblet2 vs AC Goblet"
labeled_genes <- c('REG1A', 'RPS6', 'PIGR', 'MUC5B', 'FRY', "LYZ", "LCN2", "DMBT1",
                   "CD74", "WFDC2", 'SATB2', 'HMGCS2', 'ZG16', 'CA2', 'PPARG', 'MUC4',
                   'MUC17', 'MUC2', 'MUC3A', 'TFF1')

de_results  <- markers_POU_GB2vsAC
df <- de_results[, c('avg_log2FC', 'p_val_adj')]
min_nonzero <- min(subset(df$p_val_adj, df$p_val_adj > 0))
df$p_val_adj[df$p_val_adj == 0] <- min_nonzero
colnames(df) <- c('logFC', 'madjp')
df$madjp <- 0-log10(df$madjp)

# Set significance thresholds
logFC_threshold <- 1     # Choose a log2 fold change threshold
madjp_threshold <- -log10(0.001)  # Choose an adjusted p-value threshold
df$madjp[df$madjp > 300] <- 300

# Create a new variable to color significant points
df <- df %>%
  mutate(Significant = ifelse(abs(logFC) > logFC_threshold & madjp > madjp_threshold, "Significant", "Not Significant"))

# Create a new variable for labels
df$label <- rownames(df) 
df$label[!df$label %in% labeled_genes] <- ''

p <- ggscatter(df, x = 'logFC', y = 'madjp', color = 'Significant', size = 2) +
  theme_pubr() +
  scale_color_manual(values = c("Not Significant" = "#bdbdbd", "Significant" = "#fd8d3c")) +
  theme(axis.title = element_text(size = 12, face = 'bold'),
        axis.text = element_text(size = 12),
        plot.title = element_text(hjust = 0.5),
        legend.position = "bottom")+
  labs(title="TF regulator",
       x ="Correlation To Gene Expression", y = "Max TF Motif Delta")+
  geom_text_repel(aes(label=label),
                  nudge_y = 0, max.overlaps = 1000, min.segment.length = 0.01,
                  size = 2, force = 5) +
  geom_vline(xintercept = 0, linetype = "dashed") +
  geom_hline(yintercept = madjp_threshold, linetype = "dashed") +
  geom_vline(xintercept = c(-logFC_threshold, logFC_threshold), linetype = "dashed") +
  guides(color=guide_legend(title=""))+
  labs(title = plot.title,
       x = "Log2 Fold Change",
       y = "-log10(Adjusted P-value) (capped at 300)",
       color = "Significance") +
  theme(legend.position = "bottom")
pdf(paste0(out.dir, plot.title, '.pdf'), width = 5, height = 5)
p
dev.off()
```


```{r}
plot.title = "Goblet2 vs Goblet1"
labeled_genes <- c('REG1A', 'RPS6', 'PIGR', 'MUC5B', 'FRY',
                   'MUC17', 'MUC2', 'MUC3A', 'TFF1')

de_results  <- markers_GB2vsGB1
df <- de_results[, c('avg_log2FC', 'p_val_adj')]
min_nonzero <- min(subset(df$p_val_adj, df$p_val_adj > 0))
df$p_val_adj[df$p_val_adj == 0] <- min_nonzero
colnames(df) <- c('logFC', 'madjp')
df$madjp <- 0-log10(df$madjp)

# Set significance thresholds
logFC_threshold <- 1     # Choose a log2 fold change threshold
madjp_threshold <- -log10(0.05)  # Choose an adjusted p-value threshold
df$madjp[df$madjp > 300] <- 300

# Create a new variable to color significant points
df <- df %>%
  mutate(Significant = ifelse(abs(logFC) > logFC_threshold & madjp > madjp_threshold, "Significant", "Not Significant"))

# Create a new variable for labels
df$label <- rownames(df) 
df$label[!df$label %in% labeled_genes] <- ''

p <- ggscatter(df, x = 'logFC', y = 'madjp', color = 'Significant', size = 2) +
  theme_pubr() +
  scale_color_manual(values = c("Not Significant" = "#bdbdbd", "Significant" = "#fd8d3c")) +
  theme(axis.title = element_text(size = 12, face = 'bold'),
        axis.text = element_text(size = 12),
        plot.title = element_text(hjust = 0.5),
        legend.position = "bottom")+
  labs(title="TF regulator",
       x ="Correlation To Gene Expression", y = "Max TF Motif Delta")+
  geom_text_repel(aes(label=label),
                  nudge_y = 0, max.overlaps = 1000, min.segment.length = 0.01,
                  size = 2, force = 5) +
  geom_vline(xintercept = 0, linetype = "dashed") +
  geom_hline(yintercept = madjp_threshold, linetype = "dashed") +
  geom_vline(xintercept = c(-logFC_threshold, logFC_threshold), linetype = "dashed") +
  guides(color=guide_legend(title=""))+
  labs(title = plot.title,
       x = "Log2 Fold Change",
       y = "-log10(Adjusted P-value) (capped at 300)",
       color = "Significance") +
  theme(legend.position = "bottom")
# pdf(paste0(out.dir, plot.title, '.pdf'), width = 5, height = 5)
# p
# dev.off()
```


```{r}
markers_Stem2vsStem1 <- FindMarkers(seurat, ident.1 = 'Stem2', ident.2 = 'Stem1', group.by = 'anno2', logfc.threshold = 0)
saveRDS(markers_Stem2vsStem1, paste0(out.dir, 'markers_Stem2vsStem1.rds'))

markers_Stem2vsStem1 <- readRDS(paste0(out.dir, 'markers_Stem2vsStem1.rds'))
```


```{r}
plot.title = "Stem2 vs Stem1"
labeled_genes <- c("XACT", "C10orf99", "CD24", "SATB2", "CEACAM5", "HES1", 'LEFTY1',
                   "REG1A", "DMBT1", "ALDOB", "SOX6", "SI", "CPS1" , 'LDHB', 'ADAMTSL1',
                   "HNF4G", "KRT20", "REG3A", "PLCG2", "DEFA6", 'DEFA5')

de_results  <- markers_Stem2vsStem1
df <- de_results[, c('avg_log2FC', 'p_val_adj')]
min_nonzero <- min(subset(df$p_val_adj, df$p_val_adj > 0))
df$p_val_adj[df$p_val_adj == 0] <- min_nonzero
colnames(df) <- c('logFC', 'madjp')
df$madjp <- 0-log10(df$madjp)

# Set significance thresholds
logFC_threshold <- 1     # Choose a log2 fold change threshold
madjp_threshold <- -log10(0.05)  # Choose an adjusted p-value threshold
df$madjp[df$madjp > 300] <- 300

# Create a new variable to color significant points
df <- df %>%
  mutate(Significant = ifelse(abs(logFC) > logFC_threshold & madjp > madjp_threshold, "Significant", "Not Significant"))

# Create a new variable for labels
df$label <- rownames(df) 
df$label[!df$label %in% labeled_genes] <- ''

p <- ggscatter(df, x = 'logFC', y = 'madjp', color = 'Significant', size = 2) +
  theme_pubr() +
  scale_color_manual(values = c("Not Significant" = "#bdbdbd", "Significant" = "#fd8d3c")) +
  theme(axis.title = element_text(size = 12, face = 'bold'),
        axis.text = element_text(size = 12),
        plot.title = element_text(hjust = 0.5),
        legend.position = "bottom")+
  labs(title="TF regulator",
       x ="Correlation To Gene Expression", y = "Max TF Motif Delta")+
  geom_text_repel(aes(label=label),
                  nudge_y = 0, max.overlaps = 1000, min.segment.length = 0.01,
                  size = 2, force = 5) +
  geom_vline(xintercept = 0, linetype = "dashed") +
  geom_hline(yintercept = madjp_threshold, linetype = "dashed") +
  geom_vline(xintercept = c(-logFC_threshold, logFC_threshold), linetype = "dashed") +
  guides(color=guide_legend(title=""))+
  labs(title = plot.title,
       x = "Log2 Fold Change",
       y = "-log10(Adjusted P-value) (capped at 300)",
       color = "Significance") +
  theme(legend.position = "bottom")
# pdf(paste0(out.dir, plot.title, '.pdf'), width = 5, height = 5)
# p
# dev.off()
```












